fw/http_parser.c

```diff
@@ -1474,8 +1476,10 @@ __parse_connection(TfwHttpMsg *hm, unsigned char *data, size_t len)
 	__FSM_STATE(I_ConnTok) {
 		WARN_ON_ONCE(!parser->_acc);
 
-		if (!IS_WS(c) && c != ',' && !IS_CRLF(c))
+		if (!IS_WS(c) && c != ',' && !IS_CRLF(c)) {
+			p = parser->_saved_p;
```

Comment:
It ssems the trick with `_saved_p` variable will not work it the input is divided in small chunks (say 1-2-5 bytes).

Reply:
Again, I read some older code and I find myself that I do not understand how this works.
```
#define TRY_CONN_TOKEN(name, lambda)					 \
	TRY_STR_LAMBDA_finish(name, lambda, {				 \
			if (__hbh_parser_add_data(hm, p, __fsm_n, false))\
				r = CSTR_NEQ;				 \
			else						 \
				parser->_i_st = &&I_Conn;		 \
		}, I_ConnTok)

```
This `__hbh_parser_add_data(hm, p, __fsm_n, false)` seems not to be chunk-safe, because it trying to add a chunk as a whole header.
Again, `TRY_STR_LAMBDA_finish` add parsed data, chunk-by-chung, to parser->hdr, but this data are not used. 

This stranges I found when tried to learn what the code under review should do...

Reply:
```
	if (!IS_WS(c) && c != ',' && !IS_CRLF(c)) {
```
It seems it is better to check inverted condition and set flags and jump to `I_EoT` in case.
And in this way we do not need this `_saved_p` variable at all.

Reply:
But we have the test for chunked data, right? So if there is a problem, how can we not to hit it? (And yes, I'm also confused with `__hbh_parser_add_data()` against chunked data, but still didn't dig into it.)

Reply:
Ironically, those are problems similar to ones I had with `__***_parse_cache_control()` and `__req_parse_content_type()` functions, and I also proposed to save the start of the chunk. However, we don't really need to store it in the `parser` - a local variable in `__FSM_DECLARE_VARS` is enough, because we always fixup at chunk boundaries.

parser->_acc is an ambiguous name to store the `connection` flags in, so you might consider creating union out of parser->cc_dir_flag e.g. parser->conn_flag.

Speaking about `__hbh_parser_add_data()`: in the original code `TRY_CONN_TOKEN` employed a lazy (late) chunking (fixup), which implies that calling function will fixup a remaining portion of the string - the actual amount is returned as `__data_off(p + __fsm_sz)`. However, the parent function `TFW_HTTP_PARSE_SPECHDR_VAL` fixups the string into `parser->hdr` instead of `parser->hbh_parser`, while in fact we need to append the chunk string into both.

Thus, the `TRY_CONN_TOKEN` modification seems reasonable to me. I'm wondering how the old `TRY_CONN_TOKEN` even worked. Probably we just didn't had the tests to detect the problem e.g. they tested only `data = p` cases. I fixed a similar problem with #1566, however, I did not check the correctness of `parser->hbh_parser` fixups.

With all that said, I do agree the problem can be resolved without additional `_saved_p` variable - we already effectively save the matched string and we can call `__hbh_parser_add_data()` with arbitrary `char *` constant.

Reply:
Resolve as outdated.